warehouse/cli/tuf.py

```diff
@@ -0,0 +1,206 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import json
+import os
+import tempfile
+import time
+
+from datetime import datetime
+from typing import Any, Dict, List
+
+import click
+import requests
+
+from sqlalchemy import MetaData, Table, create_engine
+from sqlalchemy.dialects.sqlite import insert
+from sqlalchemy.engine import Connection
+from tuf.api.metadata import Delegations, Metadata, Targets
+
+from warehouse.cli import warehouse
+from warehouse.packaging.tasks import (
+    generate_projects_simple_detail as _generate_projects_simple_detail,
+)
+from warehouse.packaging.utils import render_simple_detail
+
+
+@warehouse.group()  # pragma: no-branch
+def tuf():
+    """
+    Manage Warehouse's TUF state.
+    """
+
+
+@tuf.group()
+def dev():
+    """
+    TUF Development purposes commands
+    """
+
+
+def _fetch_succinct_roles(config):
```

Comment:
I think this could use some explanation. It looks like you manually tofu the first version of root in RSTUF and then use the client updater to update timestamp, snapshot and targets, to get the hash bin delegation info from the latest targets, without first updating root?

I have a few question:
- Is it unsafe to expect that top-level targets has not yet been updated when calling import_all? If not, couldn't we just use the bin numbers from `rstuf-bostrap-payload.json` to generate `Delegations` object?
- If we do need to look at the metadata, does warehouse have no other way, to get the latest metadata, than using the client updater?
- If we do need the client updater, why not use it all the way, via `refresh()`, and also update root?
- If warehouse uses the client updater and needs to bootstrap trust, can't it use the initial root metadata from `rstuf-bootstrap-payload.json`?

Reply:
> I think this could use some explanation. It looks like you manually tofu the first version of root in RSTUF and then use the client updater to update timestamp, snapshot and targets, to get the hash bin delegation info from the latest targets, without first updating root?

Let me explain the problem. I think you already gave some solutions for it.

The Warehouse user, in that case in specific, a Warehouse developer who wants to develop/test something related to RSTUF / Warehouse, will use this command to import all existent data from development Warehouse database to RSTUF database.

`warehouse tuf dev import-all` is a subcommand inside `dev` command to make clear it is for development only. (Maybe a better protection to avoid it to run in production?)

It requires Warehouse database initialization (`make initdb`) and TUF initialization (`make tufinit`) first.

What I'm trying to do here: I need to identify the Delegation role name, to fetch from RSTUF Database (table `rstuf_target_roles`) the role id when I'm inserting the artifact (package or simple detail index).

> I have a few question:
> 
> * Is it unsafe to expect that top-level targets has not yet been updated when calling import_all? If not, couldn't we just use the bin numbers from `rstuf-bostrap-payload.json` to generate `Delegations` object?`

Do we need to care about this safety here in development process?

I could use it from `rstuf-bootstrap-payload.json`, but what if I generated a custom ceremony with RSTUF CLI.
For example, as a developer user I want to test a new feature or even test performance with 16384 delegation roles and I used `kairo-payload.json` in the `Makefile` or just run the bootstrap command from `Makefile`? It could generate an inconsistency unless I change a specific call in the `cli/tuf.py`.

> * If we do need to look at the metadata, does warehouse have no other way, to get the latest metadata, than using the client updater?

I could do as I used to do before, loading the file `/var/opt/warehouse/metadata` https://github.com/kairoaraujo/warehouse/blob/20535631d6c974cc76d757a5c62dac145759960f/docker-compose.yml#L114

Another option is get it from `config.registry.settings["tuf.metadata.url"]` (http://files:9001/metadata) and get the `1.target.json?`, import it as TUF Metadata and get the succinct role.

> * If we do need the client updater, why not use it all the way, via `refresh()`, and also update root?

I assumed that no updates in the metadata and the developer did: `make initdb`, `make tufinit` and `make tufimport`  where I tried to cover only a scenario where they just  make a custome ceremony to generate a custom payload, but not a root metadata update.

> * If warehouse uses the client updater and needs to bootstrap trust, can't it use the initial root metadata from `rstuf-bootstrap-payload.json`?

I assumed yes. IMHO, We shouldn't care here about the root.

Reply:
> What I'm trying to do here: I need to identify the Delegation role name, to fetch from RSTUF Database (table `rstuf_target_roles`) the role id when I'm inserting the artifact (package or simple detail index).

Wait, why does Warehouse even need to be concerned with those technical details? Shouldn't `import-all` just throw target files at the RSTUF API, like normal release uploads would do?

MO using the tuf client here makes sense for two reasons:
1. security -- tuf client guarantees that we get the correct metadata 
2. abstraction -- tuf client hides the complexity of getting metadata

A the same time we,  we don't actually care for security here, because we ignore root updates. And we violate layering because we handle TUF metadata and manually resolve delegations inside Warehouse, which could be all hidden behind the RSTUF API.

If this is a workaround, because RSTUF does not yet implement an API endpoint for "backsigning" existing packages, and is only used in a controlled dev environment, then we should use the easiest way of getting this information, e.g. `ls /var/opt/warehouse/metadata/*.targets.json | sort -r | head -1`, or just create the object by ourselves:

```python
succinct_roles = SuccinctRoles(
   keyids,         # irrelevant
   threshold,      # irrelevant
   bit_length,     # Can't we get this from warehouse config or db?
   name_prefix,    # Can't we get this from warehouse config or db?
)
```

Reply:
Please see previous comment for how I think we should do this. Below replies are just to sort out misunderstandings...

> I could use it from `rstuf-bootstrap-payload.json`, but what if I generated a custom ceremony with RSTUF CLI. For example, as a developer user I want to test a new feature or even test performance with 16384 delegation roles and I used `kairo-payload.json` in the `Makefile` or just run the bootstrap command from `Makefile`? It could generate an inconsistency unless I change a specific call in the `cli/tuf.py`.

I didn't mean the default `rstuf-bootstrap-payload.json` included in this PR, but the one that the admin/developer actually uses to initialise RSTUF. I assumed that Warehouse would somehow keep a copy of that. 

> 
> > * If warehouse uses the client updater and needs to bootstrap trust, can't it use the initial root metadata from `rstuf-bootstrap-payload.json`?
> 
> I assumed yes. IMHO, We shouldn't care here about the root.

I didn't mean the initial root metadata that is already in RSTUF, but the one that is uploaded to RSTUF. Again I assumed that Warehouse would somehow keep a copy of that when doing `tufinit`.

Reply:
> Wait, why does Warehouse even need to be concerned with those technical details? Shouldn't `import-all` just throw target files at the RSTUF API, like normal release uploads would do?

We are doing an import, not a regular flow of adding targets using the PyPI API to add targets.

This is a procedure to speed up the process of [Importing existing targets](https://repository-service-tuf.readthedocs.io/en/latest/guide/deployment/importing-targets.html)

Unfortunately, the Warehouse doesn't support the existing tool from RSTUF due to the SQLAlchemy version, and we implemented it in the Warehouse CLI for development.

> A the same time we, we don't actually care for security here, because we ignore root updates. And we violate layering because we handle TUF metadata and manually resolve delegations inside Warehouse, which could be all hidden behind the RSTUF API.
>
> If this is a workaround, because RSTUF does not yet implement an API endpoint for "backsigning" existing packages, and is only used in a controlled dev environment, 

RSTUF Supports adding all targets to the API and publishing them to the metadata files in the storage by another API call. 
But we don't want to run this procedure (designed for high-traffic), adding many artifacts (hundred thousand to millions) at once, so we would do a kind of data migration first.

> then we should use the easiest way of getting this information, e.g. `ls /var/opt/warehouse/metadata/*.targets.json | sort -r | head -1`, or just create the object by ourselves:
> 
> ```python
> succinct_roles = SuccinctRoles(
>    keyids,         # irrelevant
>    threshold,      # irrelevant
>    bit_length,     # Can't we get this from warehouse config or db?
>    name_prefix,    # Can't we get this from warehouse config or db?
> )
> ```

Yes, I think we can go for the simpler way.

`bit_length`: We can get it from the RSTUF API
`name_prefix`:   I would say it is irrelevant, RSTUF has predefine as "bins"